flight/Modules/Actuator/actuator.c

```diff
@@ -218,279 +231,416 @@ static float get_curve2_source(ActuatorDesiredData *desired, SystemSettingsAirfr
 	return 0;
 }
 
-/**
- * @brief Main Actuator module task
- *
- * Universal matrix based mixer for VTOL, helis and fixed wing.
- * Converts desired roll,pitch,yaw and throttle to servo/ESC outputs.
- *
- * Because of how the Throttle ranges from 0 to 1, the motors should too!
- *
- * Note this code depends on the UAVObjects for the mixers being all being the same
- * and in sequence. If you change the object definition, make sure you check the code!
- *
- * @return -1 if error, 0 if success
- */
-static void actuator_task(void* parameters)
+static void compute_one_mixer(int i,
+		int16_t (*vals)[MIXERSETTINGS_MIXER1VECTOR_NUMELEM],
+		MixerSettingsMixer1TypeOptions type)
 {
-	float dT = 0.0f;
-
-	ActuatorCommandData command;
-	ActuatorDesiredData desired;
-	MixerStatusData mixerStatus;
-	FlightStatusData flightStatus;
-	ManualControlCommandData manual_control_command;
-
-	SystemSettingsAirframeTypeOptions airframe_type;
-
-	// Connect update callbacks
-	FlightStatusConnectCallbackCtx(UAVObjCbSetFlag, &flightStatusUpdated);
-	ManualControlCommandConnectCallbackCtx(UAVObjCbSetFlag, &manualControlCommandUpdated);
-
-	// Main task loop
-	uint32_t last_systime = PIOS_Thread_Systime();
+	types_mixer[i] = type;
 
-	bool rc = false;
+	i *= MIXERSETTINGS_MIXER1VECTOR_NUMELEM;
 
-	while (1) {
-		if (actuator_settings_updated) {
-			actuator_settings_updated = false;
-			ActuatorSettingsGet(&actuatorSettings);
-			actuator_set_servo_mode();
+	if ((type != MIXERSETTINGS_MIXER1TYPE_SERVO) &&
+			(type != MIXERSETTINGS_MIXER1TYPE_MOTOR)) {
+		for (int j = 0; j < MIXERSETTINGS_MIXER1VECTOR_NUMELEM; j++) {
+			// Ensure unused types are zero-filled
+			motor_mixer[i+j] = 0;
 		}
-		if (mixer_settings_updated) {
-			mixer_settings_updated = false;
-			MixerSettingsGet(&mixerSettings);
-			SystemSettingsAirframeTypeGet(&airframe_type);
+	} else {
+		for (int j = 0; j < MIXERSETTINGS_MIXER1VECTOR_NUMELEM; j++) {
+			motor_mixer[i+j] = (*vals)[j] * (1.0f / MIXER_SCALE);
 		}
+	}
+}
 
-		if (rc != true) {
-			/* Update of ActuatorDesired timed out,
-			 * or first iteration.  Go to failsafe */
-			set_failsafe();
-		}
+/* Here be dragons */
+#define compute_one_token_paste(b) compute_one_mixer(b-1, &mixerSettings.Mixer ## b ## Vector, mixerSettings.Mixer ## b ## Type)
```

Comment:
`##` is the token pasting operator; it concatenates C tokens into one bigger token.  The purpose of this macro is to token paste and remove a lot of boilerplate as a result.